### DISCUSSION TITLE
Async and refresh questions

### DISCUSSION THREAD

**Discussion by ElPiloto** [nvim-neo-tree/neo-tree.nvim#487] [Internal Development] [answered]
Hello again, I've finally gotten some time to work on another custom source and I had a few questions.

1. Is there an example of the right way to refresh my source? I normally run `:Neotree my_source` and I've been refreshing by re-running that command. But I'm wondering if there's a more official way? I'm actually not sure how `:Neotree my_source` makes its way to my source's `.navigate` function  or else I'd try to mimic that path to do a refresh more directly.

2. I'm now populating my custom source with some outputs from mercurial which can take a bit of time to load. Are there any examples of populating Neotree asynchronously so that neovim stays responsive during that long load? 

As always, thanks for your time and for making this awesome plugin.

**Comment by cseickel** [comment, marked as answer]
> 1. Is there an example of the right way to refresh my source?

Refreshing usually happens from either someone pressing `R` within the Neo-tree window to refresh manually, or automatically in response to events. You can look at the `setup` function in the `init.lua` file of any other source to see examples of subscribing to events. You can also use file watchers to refresh when a file is changed. That is more complicated and you have to make sure you unwatch folders when you are done. Only the `filesystem` source does this.

> I'm actually not sure how :Neotree my_source makes its way to my source's .navigate function or else I'd try to mimic that path to do a refresh more directly.

The commands are defined in the `plugin` directory, and they call `neo-tree.command.execute()`
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/3c70bb4a8a6cf52b534673a3e64c1213a086797f/lua/neo-tree/command/init.lua#L14-L40



> 2\. I'm now populating my custom source with some outputs from mercurial which can take a bit of time to load. Are there any examples of populating Neotree asynchronously so that neovim stays responsive during that long load?

The `filesystem` source will run asynchronously in some situations, and git status updates run asynchronously by default. Here is the code for that:
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/3c70bb4a8a6cf52b534673a3e64c1213a086797f/lua/neo-tree/git/status.lua#L203

It uses Plenary's [Job](https://github.com/nvim-lua/plenary.nvim#plenaryjob) for the async part. The git code is extra complicated because there are so many edge cases and we support showing status diffed to any arbitrary branch.
